spec/middleware_spec.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+RSpec.describe QueueBus::Middleware do
+  subject { described_class::Stack.new }
+
+  let(:args) { { rand: rand } }
+  let(:inner) { proc {} }
+
+  let(:base_class) do
+    # Creates a unique base class that provides tracking of calls.
+    Class.new(QueueBus::Middleware::Abstract) do
+      class << self
+        attr_accessor :called_at
+      end
+
+      def call(_args)
+        self.class.called_at = Time.now
+        @app.call
+      end
+    end
+  end
+
+  context 'with none configured' do
+    it 'falls through' do
+      expect(inner).to receive(:call)
+      subject.run(args, &inner)
+    end
+  end
+
+  context 'with the abstract configued' do
+    before do
+      subject.use(QueueBus::Middleware::Abstract)
+    end
+
+    it 'runs the inner' do
+      expect(inner).to receive(:call).and_call_original
+      subject.run(args, &inner)
+    end
+  end
+
+  context 'with one middleware' do
+    let(:middleware) do
+      Class.new(base_class) do
+        class << self
+          attr_accessor :args
+        end
+
+        def call(args)
+          self.class.args = args
+          super
+        end
+      end
+    end
+
+    before do
+      subject.use(middleware)
+    end
+
+    it 'calls the middleware with the args' do
+      subject.run(args, &inner)
+
+      expect(middleware.args).to eql args
+    end
+
+    it 'calls the middleware and then inner' do
+      expect(inner).to receive(:call).and_call_original
+      subject.run(args, &inner)
+      expect(middleware.called_at).not_to be_nil
+    end
+
+    context 'and the middleware does not yield' do
```

Comment:
Nice to see this test as it calls out a potentially implementation mistake when writing middleware that uses this functionality.